fn/async.go

```diff
@@ -0,0 +1,76 @@
+package fn
+
+import "errors"
+
+type Async[A any] interface {
+	Await() (*A, error)
+	Cancel()
+}
+
+var (
+	ContextCancelled = errors.New("context cancelled")
+	AsyncCancelled   = errors.New("async cancelled")
+)
+
+// RunAsyncWithCancel is a function that will transform a synchronous function
+// and make it a cancellable asynchronous call. This can be useful if the
+// function in the argument can block indefinitely and the calling context needs
+// a way to abandon waiting for it to complete.
+func RunAsyncWithCancel[A any](
+	f func() (*A, error),
+	contextCancel <-chan struct{},
+) Async[A] {
+	a := async[A]{
+		contextCancel: contextCancel,
+		targetCancel:  make(chan struct{}),
+		res:           make(chan *A),
+		err:           make(chan error),
+	}
+	go func() {
+		res, err := f()
+		if err != nil {
+			select {
+			case a.err <- err:
+			case <-a.contextCancel:
+			case <-a.targetCancel:
+			}
+		} else {
```

Comment:
nit: can avoid nested `else` by returning in the `if` block

Reply:
The crusade against `else` continues lol